Programs/_testembed.c

```diff
@@ -85,6 +85,47 @@ static int test_repeated_init_and_subinterpreters(void)
     return 0;
 }
 
+/* bpo-36225: Implicitly tear down subinterpreters with Py_Finalize() */
+static int test_finalize_subinterps(void)
+{
+    PyThreadState *mainstate;
+    PyGILState_STATE gilstate;
+    int i, j;
+
+    for (i=0; i<15; i++) {
+        printf("--- Pass %d ---\n", i);
+        _testembed_Py_Initialize();
+        mainstate = PyThreadState_Get();
+
+        PyEval_ReleaseThread(mainstate);
+
+        gilstate = PyGILState_Ensure();
+        print_subinterp();
+        PyThreadState_Swap(NULL);
+
+        for (j=0; j<2; j++) {
+            Py_NewInterpreter();
+            print_subinterp();
+        }
+
+        PyThreadState_Swap(mainstate);
+        print_subinterp();
+
+        for (j=0; j<2; j++) {
+            Py_NewInterpreter();
+            print_subinterp();
+        }
+
+        PyThreadState_Swap(mainstate);
+        print_subinterp();
+        PyGILState_Release(gilstate);
+
+        PyEval_RestoreThread(mainstate);
+        Py_Finalize();
```

Comment:
This certainly helps verify that finalization still works.  There should probably also be something verifying that the subinterpreters were properly cleaned up at the beginning of finalization.  (...perhaps with some artifact generated when each sub-interp is finalized.)

Also, what about the case where:
* the subinterpreter has multiple threads still running?
* what about daemon threads? (yeah, it's mean of me to ask :wink:)
* the subinterpreter's `tstate_head` is still running?
* someone calls `Py_NewInterpreter()` while interpreters are being cleaned up?
* someone calls `Py_NewInterpreter()` while finalization is otherwise still running?
* someone calls `Py_NewInterpreter()` after finalization is finished?

Reply:
Perhaps registering an "atexit" handler in each subinterpreter that prints something, and then confirming in the Python test case code that all the subinterpreter exit messages appear before the main interpreter's exit message?

Reply:
This all sounds worth checking in a test, but I'm unclear how to implement it. Any advice would be appreciated, specifically:
 - Should all of the test logic be in `_testembed.c`, or should that just be performing the C API calls with most of the test logic being in `test_embed.py`?
   - I'm only aware of how to register an `atexit` handler from Python code.
   - How should the interaction between `test_embed.py` and `_testembed.c` work?

With a better understanding of the above I may be able to have a go at covering the above points in the tests, although it'll likely take quite a lot of thought given I'm pretty new to the C API! Any further guidance very welcome, and I can hopefully get this finished off without such a delay this time.

Reply:
> * Should all of the test logic be in _testembed.c, or should that just be performing the C API calls with most of the test logic being in test_embed.py?

Put as much logic as you can in the Python code.  _testembed.c should mostly be only what can't be done from Python (with exceptions where practicality dictates more).

> * I'm only aware of how to register an atexit handler from Python code.

You can call Python code from C if needed.  Import the atexit module, get the appropriate function, and call it, all using the C-API.  We do the same thing in various places, like Python/import.c.  For me (not a C expert) searching the code base has always been the easiest way to see how to do something. :smile:

(That assumes there isn't a C-API for atexit handlers.)

> * How should the interaction between test_embed.py and _testembed.c work?

I'll need more context.

Reply:
This might be a good reason to pair up on a video call.  Then we could walk through this stuff a bit more efficiently.  What do you think?

Reply:
At least some of the embedding tests already use `PyRun_SimpleString()` to run Python code inside the created interpreter, and that's also what I had in mind for the suggested atexit test case above.

Reply:
It looks like there's a lot of considerations and things to check here, also fleshed out by Victor's message at [bpo-36225#msg371571](https://bugs.python.org/issue36225#msg371571). Does everything here need addressing in this one PR, or can some of these points be split into separate issues to follow this fix? This feels like rather a lot to tackle all in one go - which of the cases you listed would you suggest starting with @ericsnowcurrently (perhaps the simplest to test!)?